src/compute/src/sink/s3_oneshot.rs

```diff
@@ -27,18 +32,108 @@ where
+        // An encapsulation of the copy to response protocol.
+        // Used to send rows and errors if this fails.
+        let response_protocol_handle = Rc::new(RefCell::new(Some(ResponseProtocol {
+            sink_id,
+            response_buffer: Some(Rc::clone(&compute_state.copy_to_response_buffer)),
+        })));
+        let response_protocol_weak = Rc::downgrade(&response_protocol_handle);
+
+        copy_to(
+            sinked_collection,
+            err_collection,
+            sink_id,
+            sink.up_to.clone(),
+            response_protocol_handle,
+        );
+
+        Some(Rc::new(scopeguard::guard((), move |_| {
+            if let Some(protocol_handle) = response_protocol_weak.upgrade() {
+                std::mem::drop(protocol_handle.borrow_mut().take())
+            }
+        })))
+    }
+}
+
+fn copy_to<G>(
+    sinked_collection: Collection<G, Row, Diff>,
+    err_collection: Collection<G, DataflowError, Diff>,
+    sink_id: GlobalId,
+    up_to: Antichain<G::Timestamp>,
+    response_protocol_handle: Rc<RefCell<Option<ResponseProtocol>>>,
+) where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let name = format!("s3-oneshot-{}", sink_id);
+    let mut op = OperatorBuilder::new(name, sinked_collection.scope());
+    let mut ok_input = op.new_input(&sinked_collection.inner, Pipeline);
+    let mut err_input = op.new_input(&err_collection.inner, Pipeline);
+
+    op.build(|_cap| {
+        let mut row_count: u64 = 0;
+        let mut finished = false;
+        let mut ok_buf = Default::default();
+
+        move |frontiers| {
+            if finished {
+                // Drain the inputs, to avoid the operator being constantly rescheduled
+                ok_input.for_each(|_, _| {});
+                err_input.for_each(|_, _| {});
+                return;
+            }
+
+            let mut frontier = Antichain::new();
+            for input_frontier in frontiers {
+                frontier.extend(input_frontier.frontier().iter().copied());
+            }
+
+            ok_input.for_each(|_, data| {
+                data.swap(&mut ok_buf);
+                let count: u64 = ok_buf.drain(..).len().try_into().expect("usize into u64");
+                // TODO (mouli): write these to s3 in subsequent PRs
+                row_count = row_count + count;
+            });
+            // TODO (mouli): handle errors
+            if PartialOrder::less_equal(&up_to, &frontier) {
+                finished = true;
+                // We are done, send the final count.
+                if let Some(response_protocol) = response_protocol_handle.borrow_mut().deref_mut() {
+                    response_protocol.send_final_count(row_count);
+                }
+            }
+        }
+    });
+}
+
+/// A type that guides the transmission of number of rows back to the coordinator.
+struct ResponseProtocol {
+    pub sink_id: GlobalId,
+    pub response_buffer: Option<Rc<RefCell<Vec<(GlobalId, CopyToResponse)>>>>,
+}
```

Comment:
This should implement `Drop` to ensure a `Dropped` response is sent if (and only if) no other response was sent previously.

Reply:
Added! Also, since the handle to the buffer is cleared after sending one message, the `Dropped` response will not be sent if a message was sent previously.